clarity/src/vm/ast/parser_v2/mod.rs

```diff
@@ -0,0 +1,2518 @@
+            }
+            Token::Utf8String(data) => {
+                let str_val =
+                    Value::Sequence(SequenceData::String(CharType::UTF8(UTF8Data { data })));
+                let mut e = PreSymbolicExpression::atom_value(str_val);
+                e.span = token.span;
+                Some(e)
+            }
+            Token::Ident(mut name) => {
+                if name.len() > MAX_STRING_LEN {
+                    self.add_diagnostic(
+                        ParseErrors::NameTooLong(name.clone()),
+                        token.span.clone(),
+                    )?;
+                    // Continue with a placeholder name
+                    name = "placeholder".to_string();
+                }
+                let cname = match ClarityName::try_from(name.clone()) {
+                    Ok(name) => name,
+                    Err(_) => {
+                        self.add_diagnostic(
+                            ParseErrors::IllegalClarityName(name.clone()),
+                            token.span.clone(),
+                        )?;
+                        ClarityName::try_from("placeholder").unwrap()
+                    }
+                };
+                let mut e = PreSymbolicExpression::atom(cname);
+                e.span = token.span;
+                Some(e)
+            }
+            Token::TraitIdent(mut name) => {
+                if name.len() > MAX_STRING_LEN {
+                    self.add_diagnostic(
+                        ParseErrors::NameTooLong(name.clone()),
+                        token.span.clone(),
+                    )?;
+                    // Continue with a placeholder name
+                    name = "placeholder".to_string();
+                }
+                let cname = match ClarityName::try_from(name.clone()) {
+                    Ok(name) => name,
+                    Err(_) => {
+                        self.add_diagnostic(
+                            ParseErrors::IllegalTraitName(name.clone()),
+                            token.span.clone(),
+                        )?;
+                        ClarityName::try_from("placeholder").unwrap()
+                    }
+                };
+                let mut e = PreSymbolicExpression::trait_reference(cname);
+                e.span = token.span;
+                Some(e)
+            }
+            Token::Bytes(data) => {
+                let value = match Value::buff_from(data) {
+                    Ok(value) => value,
+                    _ => {
+                        self.add_diagnostic(ParseErrors::InvalidBuffer, token.span.clone())?;
+                        Value::buff_from_byte(0)
+                    }
+                };
+                let mut e = PreSymbolicExpression::atom_value(value);
+                e.span = token.span;
+                Some(e)
+            }
+            Token::Principal(addr) => {
+                let principal = match PrincipalData::parse_standard_principal(&addr) {
```

Comment:
I see that this is where the check that the `addr` is a well-formed address comes in.  However, if the user creates a principal literal that includes an invalid Crockford 32 character, they'll get `ParseErrors::InvalidPrincipalLiteral` instead of `LexerError::InvalidCharPrincipal`.  Is this the desired behavior?

Reply:
Updated in 2d825c5 to now generate the `InvalidCharPrincipal`.